### ISSUE TITLE
Abort when `error_already_set` is thrown inside `PYBIND11_PLUGIN`

### ISSUE THREAD

**Issue by TallJimbo** [pybind/pybind11#686]
When `error_already_set` is thrown by Python C++ API code inside the `PYBIND11_PLUGIN` macro block, the `std::exception` catch block defined by the macro somehow doesn't work, and the result is a `py::error_already_set` that propagates up and terminates the program, but with a *different* Python exception type and message than the original; regardless of what the original is, I always see:
```
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  SystemError: <built-in method __contains__ of dict object at 0x7ffff6bab480> returned a result with an error set
```

I haven't been able to figure out how this is happening, but it's easily fixed by adding an extra
```
try {
    ...
} catch (pybind11::error_already_set & err) {
    err.restore();
    return nullptr;
}
```
around the offending code.  My best guess is that the `error_already_set` destructor is causing another instance of `py::error_already_set` to be raised somehow, but that too seems impossible.

In any case, my recommendation is that we add an explicit catch block for `error_already_set` to the `PYBIND11_PLUGIN` macro that simply uses that exception rather than raising `ImportError` (or at least uses the original error message).  That will make it much easier to diagnose failing code.

Here's a complete example to reproduce the problem:
```
#include "pybind11/pybind11.h"

PYBIND11_PLUGIN(import_abort) {
    pybind11::module mod("import_abort");
    pybind11::module::import("some-other-module-that-does-not-exist");
    return mod.ptr();
}
```
Simply import this module to trigger the failure.

**Comment by dean0x7d** [comment]
Have you tried this with the latest code on the master branch? I believe this was already fixed in #657.

**Comment by TallJimbo** [comment]
That explains it; I have not.  Sorry for the noise.
